src/cracker.c

```diff
@@ -1060,10 +1071,14 @@ static int crk_salt_loop(void)
 	crk_process_key_max_keys = 0; /* use slow path next time */
 	crk_key_index = 0;
 	crk_last_salt = NULL;
-	if (options.flags & FLG_MASK_STACKED)
-		mask_fix_state();
-	else
-	crk_fix_state();
+
+	if (fix_state_present && event_fix_state) {
+		if (options.flags & FLG_MASK_STACKED)
+			mask_fix_state();
+		else if (crk_fix_state)
+			crk_fix_state();
+		event_fix_state = 0;
+	}
```

Comment:
Maybe instead of deciding on which function to call each time, we should conditionally set the function pointer to `mask_fix_state()` in `crk_init()`. Also, why do we need `else if` here (rather than just `else`) - is it somehow possible that we have `fix_state_present` yet aren't going to call any function here?

Reply:
Right! I somehow overlooked the former (to set the function pointer). The latter (`fix_state_present` vs `if (crk_fix_state)`) is actually an artefact. I will amend it, thanks!

Reply:
Here's the deal: `mask_fix_state()` will sometimes call `crk_fix_state()` so we can't change that. Hybrid "fix state" stuff are very competent foot guns, I have plans to revisit the strange `hybrid_fix_state` than Jim added in the weirdest place.

Anyway I will drop the `else if`.